File: grindx/judges/cpp/maximum-depth-of-binary-tree.cpp
/**
 * Judge for Maximum Depth of Binary Tree — function pattern, tree input, int return.
 */

#include "_common.h"
#include "solution.cpp"

#include <random>

int main() {
    auto tc = load_cases("maximum-depth-of-binary-tree");
    auto& cases = tc["cases"].get_array();

    // Generate large case for TLE detection: a complete binary tree with ~10000 nodes
    struct LargeCase {
        json::Array tree_arr;
        int expected_depth;
    };
    std::vector<LargeCase> large_cases;
    {
        // Build a complete binary tree of depth 14 (2^14 - 1 = 16383 nodes)
        int depth = 14;
        int n = (1 << depth) - 1;
        json::Array arr;
        std::mt19937 rng(42);
        std::uniform_int_distribution<int> dist(-100000, 100000);
        for (int k = 0; k < n; k++) arr.push_back(json::Value(dist(rng)));
        large_cases.push_back({arr, depth});
    }

    int total = static_cast<int>(cases.size()) + static_cast<int>(large_cases.size());

    // Basic cases
    for (int i = 0; i < static_cast<int>(cases.size()); i++) {
        auto& c = cases[i];
        auto& arr = c["input"][0].get_array();
        TreeNode* root = list_to_tree(arr);

        int result = maxDepth(root);
        int expected = c["expected"].get<int>();

        if (result != expected) {
            std::string cat = c["category"].is_string() ? c["category"].get_string() : "";
            report_wa(i, c["input"].to_string(), std::to_string(expected),
                std::to_string(result), total, cat);
        }
    }

    // Large cases
    for (int li = 0; li < static_cast<int>(large_cases.size()); li++) {
        int idx = static_cast<int>(cases.size()) + li;
        auto& lc = large_cases[li];
        TreeNode* root = list_to_tree(lc.tree_arr);

        int result = maxDepth(root);

        if (result != lc.expected_depth) {
            report_wa(idx, "large input (16383 nodes)", std::to_string(lc.expected_depth),
                std::to_string(result), total, "tle");
        }
    }

    report_ac(total);
}

File: grindx/judges/java/maximum-depth-of-binary-tree.java
import java.util.*;

/**
 * Judge for Maximum Depth of Binary Tree — function pattern, tree input, int return.
 *
 * NOTE: The user's Solution.java defines a top-level TreeNode class, which is
 * a different type from Common.TreeNode. Therefore, tree deserialization is
 * inlined here using the user's TreeNode directly.
 */
class Judge {

    /** Build a binary tree from a level-order list (with nulls) using the user's top-level TreeNode. */
    static TreeNode listToTree(List<Object> arr) {
        if (arr == null || arr.isEmpty() || arr.get(0) == null) return null;
        TreeNode root = new TreeNode(((Number) arr.get(0)).intValue());
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        int i = 1;
        while (i < arr.size() && !queue.isEmpty()) {
            TreeNode node = queue.poll();
            // Left child
            if (i < arr.size()) {
                Object leftVal = arr.get(i);
                i++;
                if (leftVal != null) {
                    node.left = new TreeNode(((Number) leftVal).intValue());
                    queue.add(node.left);
                }
            }
            // Right child
            if (i < arr.size()) {
                Object rightVal = arr.get(i);
                i++;
                if (rightVal != null) {
                    node.right = new TreeNode(((Number) rightVal).intValue());
                    queue.add(node.right);
                }
            }
        }
        return root;
    }

    @SuppressWarnings("unchecked")
    public static void main(String[] args) {
        Map<String, Object> tc = Common.loadCases("maximum-depth-of-binary-tree");
        List<Object> cases = (List<Object>) tc.get("cases");

        // Generate large case for TLE detection: complete binary tree of depth 14
        List<Object[]> largeCases = generateLargeCases();
        int total = cases.size() + largeCases.size();

        // Basic cases
        for (int i = 0; i < cases.size(); i++) {
            Map<String, Object> c = (Map<String, Object>) cases.get(i);
            List<Object> input = (List<Object>) c.get("input");
            List<Object> treeArr = (List<Object>) input.get(0);

            TreeNode root = listToTree(treeArr);
            Solution sol = new Solution();
            int result = sol.maxDepth(root);

            int expected = Common.toInt(c.get("expected"));

            if (result != expected) {
                String category = c.containsKey("category") ? (String) c.get("category") : "";
                Common.reportWA(i, input, String.valueOf(expected),
                    String.valueOf(result), total, category);
            }
        }

        // Large cases
        for (int li = 0; li < largeCases.size(); li++) {
            int idx = cases.size() + li;
            Object[] lc = largeCases.get(li);
            List<Object> treeArr = (List<Object>) lc[0];
            int expectedDepth = (int) lc[1];

            TreeNode root = listToTree(treeArr);
            Solution sol = new Solution();
            int result = sol.maxDepth(root);

            if (result != expectedDepth) {
                Common.reportWA(idx, "large input (" + treeArr.size() + " nodes)",
                    String.valueOf(expectedDepth), String.valueOf(result), total, "tle");
            }
        }

        Common.reportAC(total);
    }

    @SuppressWarnings("unchecked")
    private static List<Object[]> generateLargeCases() {
        List<Object[]> cases = new ArrayList<>();
        Random rng = new Random(42);
        int depth = 14;
        int n = (1 << depth) - 1; // 16383 nodes
        List<Object> arr = new ArrayList<>();
        for (int k = 0; k < n; k++) {
            arr.add(rng.nextInt(200001) - 100000);
        }
        cases.add(new Object[]{arr, depth});
        return cases;
    }
}
